internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -729,3 +837,79 @@ func (a *Adapter) updateComponentSource(ctx context.Context, c client.Client, co
 	}
 	return nil
 }
+
+func (a *Adapter) prepareGroupSnapshot(application *applicationapiv1alpha1.Application, prGroupHash string) (*applicationapiv1alpha1.Snapshot, []gitops.ComponentSnapshotInfo, error) {
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.context, a.client, application)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	snapshotComponents := make([]applicationapiv1alpha1.SnapshotComponent, 0)
+	componentSnapshotInfos := make([]gitops.ComponentSnapshotInfo, 0)
+	for _, applicationComponent := range *applicationComponents {
+		applicationComponent := applicationComponent // G601
+		snapshots, err := a.loader.GetMatchingComponentSnapshotsForComponentAndPRGroupHash(a.context, a.client, a.snapshot, applicationComponent.Name, prGroupHash)
+		if err != nil {
+			a.logger.Error(err, "Failed to fetch Snapshots for component", "component.Name", applicationComponent.Name)
+			return nil, nil, err
+		}
+
+		if len(*snapshots) > 0 {
+			sortedSnapshots := gitops.SortSnapshots(*snapshots)
+
+			// find the latest component snapshot created for the open PR/MR
+			for _, snapshot := range sortedSnapshots {
+				snapshot := snapshot
+				// find the built image for pull/merge request build PLR from the latest opened pull request component snapshot
+				isOpened, err := isPRMRInSnapshotOpened(a.context, a.client, &snapshot)
+				if err != nil {
+					return nil, nil, err
+				}
+				if isOpened {
+					matchingSnapshotComponent := gitops.FindMatchingSnapshotComponent(&snapshot, applicationComponent)
+					snapshotComponents = append(snapshotComponents, matchingSnapshotComponent)
+					componentSnapshotInfos = append(componentSnapshotInfos, gitops.ComponentSnapshotInfo{
+						Component:        applicationComponent.Name,
+						BuildPipelineRun: snapshot.Labels[gitops.BuildPipelineRunNameLabel],
+						Snapshot:         snapshot.Name,
+					})
+				}
+			}
+		}
+	}
+
+	if len(snapshotComponents) == 0 {
```

Comment:
Related to the other comments - the group Snapshot should always contain all components of the application even if they weren't updated as part of the PR group - in that case they would just reference the GCL images.

Because of that maybe this check would need to check the `GroupSnapshotInfoAnnotation` to check if there are zero entries in it?

Reply:
I updated to check if `len(componentSnapshotInfos) < 2` here and it can be reviewed now. Thanks.